File: neet_code_all_cpp/array_hash/repeated_dna_sequences.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

class Solution {
    public:
        // AC
        std::vector<std::string> findRepeatedDnaSequences(std::string s) {
            if(s.size() < 10) {
                return std::vector<std::string>{};
            }

            auto n = (int) s.size();
            std::unordered_map<std::string, int> map;
            for(int i = 0; i <= n - 10; ++i) {
                ++map[s.substr(i, 10)];
            }

            std::vector<std::string> result;
            for(auto& [sub_str, n]: map) {
                if(n > 1) {
                    result.push_back(sub_str);
                }
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        std::vector<std::string> findRepeatedDnaSequences(std::string s) {
            auto n = (int)s.size(); 
            if(n <= 10) {
                return {};
            }

            std::vector<std::string> result;
            std::unordered_map<std::string, int> map;

            for(int i = 0; i <= n - 10; ++i) {
                auto sub_str = s.substr(i, 10);
                ++map[sub_str];
                if(map[sub_str] == 2) {
                    result.push_back(sub_str);
                }
            }

            return result;
        }
};

void show_result(std::vector<std::string>& result) {
    for(auto& v: result) {
        std::cout << v << " ";
    }
    std::cout << std::endl;
}

int main(void) {
    std::string case_1 = "AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT";
    // => ["AAAAACCCCC", "CCCCCAAAAA"]
    std::string case_2 = "AAAAAAAAAAAAA";
    // => ["AAAAAAAAAAAAA"]
    std::string case_3 = "AAAAAAAAAAA";
    // => ["AAAAAAAAAA"]

    Solution s_1;
    auto res_1 = s_1.findRepeatedDnaSequences(case_1);
    auto res_2 = s_1.findRepeatedDnaSequences(case_2);
    auto res_3 = s_1.findRepeatedDnaSequences(case_3);

    show_result(res_1);
    show_result(res_2);
    show_result(res_3);

    SolutionAns s_ans;
    auto res_1_ans = s_ans.findRepeatedDnaSequences(case_1);
    auto res_2_ans = s_ans.findRepeatedDnaSequences(case_2);
    auto res_3_ans = s_ans.findRepeatedDnaSequences(case_3);

    show_result(res_1_ans);
    show_result(res_2_ans);
    show_result(res_3_ans);
}

File: neet_code_all/src/array_hash/repeated_dna_sequences.rs
use std::collections::HashMap;

struct Solution {}
impl Solution {
    // AC
    pub fn find_rpeated_dna_sequences(s: String) -> Vec<String> {
        if s.len() < 10 {
            return vec![];
        }

        let n = s.len();
        let mut map: HashMap<String, i32> = HashMap::new();

        for i in 0..=n - 10 {
            //println!("i,sub_str: {i}, {}", s[i..(i + 10)].to_string());
            *map.entry(s[i..(i + 10)].to_string().clone()).or_default() += 1;
        }
        //println!("map: {:?}", map);

        let mut result = vec![];
        for (sub_seq, n) in map {
            if n > 1 {
                result.push(sub_seq);
            }
        }

        result
    }
}

// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    // AC
    pub fn find_rpeated_dna_sequences(s: String) -> Vec<String> {
        if s.len() <= 10 {
            return vec![];
        }

        let n = s.len();
        let mut map: HashMap<String, i32> = HashMap::new();
        let mut result = vec![];

        for i in 0..=(n - 10) {
            let sub_str = s[i..(i + 10)].to_string();
            *map.entry(sub_str.clone()).or_default() += 1;
            if let Some(n_sub_str) = map.get(&sub_str) {
                if *n_sub_str == 2 {
                    result.push(sub_str);
                }
            }
        }

        result
    }
}

fn main() {
    let case_1 = "AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT".to_string();
    // => ["AAAAACCCCC", "CCCCCAAAAA"]
    let case_2 = "AAAAAAAAAAAAA".to_string();
    // => ["AAAAAAAAAAAAA"]
    let case_3 = "AAAAAAAAAAA".to_string();
    // => ["AAAAAAAAAA"]

    println!(
        "case_1: {:?}",
        Solution::find_rpeated_dna_sequences(case_1.clone())
    );
    println!(
        "case_2: {:?}",
        Solution::find_rpeated_dna_sequences(case_2.clone())
    );
    println!(
        "case_3: {:?}",
        Solution::find_rpeated_dna_sequences(case_3.clone())
    );

    println!(
        "case_1: {:?}",
        SolutionAnsCpp::find_rpeated_dna_sequences(case_1.clone())
    );
    println!(
        "case_2: {:?}",
        SolutionAnsCpp::find_rpeated_dna_sequences(case_2.clone())
    );
    println!(
        "case_3: {:?}",
        SolutionAnsCpp::find_rpeated_dna_sequences(case_3.clone())
    );
}
